lib/Fencer/Rules.hs

```diff
@@ -41,11 +65,40 @@ loadRulesFromDirectory
     do
     let directory = rootDirectory </> subDirectory
     files <- listAllFiles directory
-    mapM Yaml.decodeFileThrow $
-        if ignoreDotFiles
-            then filter (not . isDotFile) files
-            else files
+    let filteredFiles = if ignoreDotFiles
+        then filter (not . isDotFile) files
+        else files
+    foldl'
```

Comment:
Actually, since it turns out you're not using the Applicative instance of `Validation` anyway, it's questionable whether we need it – the only benefit we're getting is `Success` and `Failure` instead of `Left` and `Right`, but we're paying for it with an extra dependency.

So perhaps let's just use `Either` instead.

Reply:
As for `combine`, it does a bit too much.

What about this?

(See [`partitionEithers`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Either.html#v:partitionEithers).)

```haskell
loadRulesFromDirectory = do
    ...
    (errs, rules) <- partitionEithers <$> mapM loadFile filteredFiles
    pure $ if null errs then Right rules else Left errs
  where
    loadFile :: FilePath -> IO (Either LoadRulesError DomainDefinition)
    loadFile = ...
```

Reply:
> Actually, since it turns out you're not using the Applicative instance of `Validation` anyway, it's questionable whether we need it – the only benefit we're getting is `Success` and `Failure` instead of `Left` and `Right`, but we're paying for it with an extra dependency.
> 
> So perhaps let's just use `Either` instead.

That's a good point. My motivation for using the `validation` library was in accumulating errors on the left. However, it turned out that the way two `Validation` values get combined is not what I needed, hence I ended up writing my own version of it, as witnessed by the helper function `merge`.

I'm reverting this to `Either`.

Reply:
> As for `combine`, it does a bit too much.
> 
> What about this?
> 
> (See [`partitionEithers`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Either.html#v:partitionEithers).)
> 
> ```haskell
> loadRulesFromDirectory = do
>     ...
>     (errs, rules) <- partitionEithers <$> mapM loadFile filteredFiles
>     pure $ if null errs then Right rules else Left errs
>   where
>     loadFile :: FilePath -> IO (Either LoadRulesError DomainDefinition)
>     loadFile = ...
> ```

I like your version much more! It's more succinct and easier to understand. I adopted it.